File: Dynamic_Programming/3333. Find the Original Typed String II.cpp
/*
3333. Find the Original Typed String II
*/

// TC:O(n + m*k) m 為組數 SC:O(m*k + n)
class Solution {
    using LL = long long;
    LL MOD = 1e9+7;
public:
    int possibleStringCount(string word, int k) {
        int n = word.size();
        if(n < k) return 0;
        vector<LL> cnts;
        LL res = 1;
        LL count = 0;
        for(int i = 0; i < n; i++) {
            count += 1;
            if(i == n-1 || word[i] != word[i+1]) {
                if(cnts.size() < k) {
                    cnts.push_back(count);
                }
                res = (res * count) % MOD;
                count = 0;
            }
        }

        int m = cnts.size();
        if(m >= k) return res;

        vector<vector<LL>> dp(m+1, vector<LL>(k)); // 走到某個位置時 長度小於j的數量
        dp[0][0] = 1;
        vector<LL> presum(k+1);
        for(int i = 0; i < m; i++) {
            for(int j = 0; j < k; j++) {
                presum[j+1] = (presum[j] + dp[i][j]) % MOD;
            }

            // 前面肯定已經有i個了 所以現在至少i+1
            for(int j = i + 1; j < k; j++) {
                dp[i+1][j] = (presum[j] - presum[max(j-cnts[i], 0LL)]) % MOD; 
            }
        }

        res -= accumulate(dp[m].begin() + m, dp[m].end(), 0LL); // 因為之前加總的答案不會有一個字元沒選到的狀況，所以m組至少要從長度m開始計算

        return (res % MOD + MOD) % MOD;
    }
};

File: Dynamic_Programming/3333. Find the Original Typed String II.py
# TC:O(n + k^2) SC:O(k^2)
class Solution:
    def possibleStringCount(self, word: str, k: int) -> int:
        '''
        字串至少要有k長度 問說他可以犯錯0 ~ 很多次 能有幾種可能
        n = 1e5
        k = 2000
        正難則反
        1. n < k 沒方案
        '''
        MOD = 10 ** 9 + 7
        n = len(word)
        if n < k:
            return 0
        cnts = []
        cnt = 0
        ans = 1
        for i in range(n):
            cnt += 1
            if i == n - 1 or word[i] != word[i + 1]:
                if cnt > 1:
                    if k > 0:
                        cnts.append(cnt - 1) # 每個物品都選了一個的狀況下 還可以選多少個
                    ans = ans * cnt % MOD
                k -= 1
                cnt = 0
        if k <= 0:
            return ans # 隨便選都合法
        dp = [[0] * k for _ in range(len(cnts) + 1)]
        dp[0] = [1] * k # 都不選也是一種方案
        for i, c in enumerate(cnts):
            s = list(accumulate(dp[i], initial = 0))
            for j in range(k):
                dp[i + 1][j] = (dp[i + 1][j] + s[j + 1] - s[max(0, j - c)]) % MOD
        return (ans - dp[-1][-1]) % MOD
                
